src/cwe_checker_lib/src/analysis/taint/state.rs

```diff
@@ -202,86 +212,79 @@ impl State {
         }
     }
 
-    /// Return true if the memory object with the given ID contains a tainted value.
+    /// Returns true if the memory object with the given ID contains a tainted
+    /// value.
     pub fn check_mem_id_for_taint(&self, id: &AbstractIdentifier) -> bool {
-        if let Some(mem_object) = self.memory_taint.get(id) {
-            for elem in mem_object.values() {
-                if elem.is_tainted() {
-                    return true;
-                }
-            }
-        }
-        false
+        self.memory_taint
+            .get(id)
+            .is_some_and(|mem_object| mem_object.values().any(|elem| elem.is_tainted()))
     }
 
-    /// If the given address points to the stack,
-    /// return true if and only if the value at that stack position is tainted.
-    /// If the given address points to a non-stack memory object,
-    /// return true if the memory object contains any tainted value (at any position).
-    pub fn check_if_address_points_to_taint(
-        &self,
-        address: Data,
-        pi_state: &PointerInferenceState,
-    ) -> bool {
-        use crate::analysis::pointer_inference::object::ObjectType;
-        for (target, offset) in address.get_relative_values() {
-            if let Ok(Some(ObjectType::Stack)) = pi_state.memory.get_object_type(target) {
-                // Only check if the value at the address is tainted
+    /// Check if the given address points to tainted memory.
+    ///
+    /// Returns true iff the value at any of the exact memory locations that the
+    /// pointer may point to is tainted.
+    // FIXME: The original implementation in CWE476 distinguished between stack
+    // memory and all other memory objects. If the given address points to a
+    // non-stack memory object, it returns true if the memory object contains
+    // any tainted value (at any position).
+    pub fn check_if_address_points_to_taint(&self, address: PiData) -> bool {
+        address
+            .get_relative_values()
+            .iter()
+            .any(|(target, offset)| {
                 if let (Some(mem_object), Ok(target_offset)) =
                     (self.memory_taint.get(target), offset.try_to_bitvec())
                 {
-                    if let Some(taint) = mem_object.get_unsized(target_offset.clone()) {
-                        if taint.is_tainted() {
-                            return true;
-                        }
-                    }
-                }
-            } else {
-                // Check whether the memory object contains any taint.
-                if self.check_mem_id_for_taint(target) {
-                    return true;
+                    mem_object
+                        .get_unsized(target_offset.clone())
+                        .is_some_and(|taint| taint.is_tainted())
+                } else {
+                    false
                 }
-            }
-        }
-
-        false
+            })
```

Comment:
You are right, we should keep it because it uses `get_unsized`. And then I think we do not need a size parameter, because this use case is already covered by the `load_taint_from_memory` method.